server/sonar-server/src/main/java/org/sonar/server/computation/step/PersistTestsStep.java

```diff
@@ -0,0 +1,193 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.computation.step;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ListMultimap;
+import org.sonar.api.resources.Qualifiers;
+import org.sonar.api.utils.System2;
+import org.sonar.batch.protocol.output.BatchReport;
+import org.sonar.batch.protocol.output.BatchReportReader;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.persistence.MyBatis;
+import org.sonar.core.source.db.FileSourceDto;
+import org.sonar.core.source.db.FileSourceDto.Type;
+import org.sonar.server.computation.ComputationContext;
+import org.sonar.server.computation.source.ReportIterator;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.source.db.FileSourceDb;
+
+import javax.annotation.CheckForNull;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+public class PersistTestsStep implements ComputationStep {
+
+  private final DbClient dbClient;
+  private final System2 system;
+
+  public PersistTestsStep(DbClient dbClient, System2 system) {
+    this.dbClient = dbClient;
+    this.system = system;
+  }
+
+  @Override
+  public String[] supportedProjectQualifiers() {
+    return new String[] {Qualifiers.PROJECT};
+  }
+
+  @Override
+  public void execute(ComputationContext computationContext) {
+    DbSession session = dbClient.openSession(false);
+    try {
+      int rootComponentRef = computationContext.getReportMetadata().getRootComponentRef();
+      TestContext context = new TestContext(session, computationContext);
+      recursivelyProcessComponent(context, rootComponentRef);
+      session.commit();
+    } finally {
+      MyBatis.closeQuietly(session);
+    }
+  }
+
+  private void recursivelyProcessComponent(TestContext context, int componentRef) {
+    BatchReportReader reportReader = context.reader;
+    BatchReport.Component component = reportReader.readComponent(componentRef);
+    if (component.getIsTest() && reportReader.readTests(componentRef) != null) {
+      persistTestResults(component, context);
+    }
+
+    for (Integer childRef : component.getChildRefList()) {
+      recursivelyProcessComponent(context, childRef);
+    }
+  }
+
+  private void persistTestResults(BatchReport.Component component, TestContext context) {
+    ReportIterator<BatchReport.Test> testIterator = new ReportIterator<>(context.reader.readTests(component.getRef()), BatchReport.Test.PARSER);
+    ListMultimap<String, FileSourceDb.Test.CoveredFile> nameToCoveredFiles = loadCoverageDetails(component.getRef(), context);
+    List<FileSourceDb.Test> tests = new ArrayList<>();
+
+    while (testIterator.hasNext()) {
+      BatchReport.Test batchTest = testIterator.next();
```

Comment:
Could you please move followings lines in another method ? It will makes it more readable